File: src/lib/subscription/subscription.ts
import "server-only";

import { auth } from "@clerk/nextjs/server";
import { redirect } from "next/navigation";
import { db } from "@/lib/db";
import { eq } from "drizzle-orm";
import { subscriptions } from "@/lib/db/schema";
import { Subscription, SubscriptionPlan } from "@/types/database";
import { getCurrentUser } from "../user/user";
import { routes } from "@/lib/routes";

const subscriptionPlanLevel: Record<SubscriptionPlan, number> = {
  basic: 1,
  premium: 2,
  all_in: 3,
};

export const getSubscriptionByUserId = async (
  userId: number,
): Promise<Subscription | null> => {
  const subscription = await db.query.subscriptions.findFirst({
    where: eq(subscriptions.userId, userId),
  });

  return subscription ?? null;
};

export const getCurrentUserSubscription =
  async (): Promise<Subscription | null> => {
    const user = await getCurrentUser();

    if (!user) {
      return null;
    }

    return getSubscriptionByUserId(user.id);
  };

export const isSubscriptionActive = (
  subscription: Subscription | null,
): subscription is Subscription => {
  if (!subscription) {
    return false;
  }

  return ["active", "trialing"].includes(subscription.status);
};

export const canAccessSubscriptionPlan = (
  subscription: Subscription | null,
  requiredPlan: SubscriptionPlan,
): boolean => {
  if (!isSubscriptionActive(subscription)) {
    return false;
  }

  return (
    subscriptionPlanLevel[subscription.plan] >=
    subscriptionPlanLevel[requiredPlan]
  );
};

export const canCurrentUserAccessSubscriptionPlan = async (
  requiredPlan: SubscriptionPlan,
): Promise<boolean> => {
  const subscription = await getCurrentUserSubscription();

  return canAccessSubscriptionPlan(subscription, requiredPlan);
};

export const requireCurrentUserSubscriptionPlan = async (
  requiredPlan: SubscriptionPlan,
): Promise<void> => {
  const { userId } = await auth();

  if (!userId) {
    redirect(routes.signIn);
  }

  if (!(await canCurrentUserAccessSubscriptionPlan(requiredPlan))) {
    redirect(routes.pricing);
  }
};

File: backend/auth/subscription.py
from fastapi import HTTPException
from sqlalchemy import select
from starlette import status

from auth.auth import current_user_dependency
from db.database import db_dependency
from db.models import Subscription, SubscriptionPlan

SUBSCRIPTION_PLAN_LEVELS = {
    SubscriptionPlan.BASIC: 1,
    SubscriptionPlan.PREMIUM: 2,
    SubscriptionPlan.ALL_IN: 3,
}
ACTIVE_SUBSCRIPTION_STATUSES = {"active", "trialing"}


def require_subscription_plan(required_plan: SubscriptionPlan):
    def dependency(
        db: db_dependency,
        user: current_user_dependency,
    ) -> Subscription:
        subscription = db.scalar(
            select(Subscription).where(Subscription.user_id == user.id)
        )

        if subscription is None:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="An active subscription is required",
            )

        if subscription.status not in ACTIVE_SUBSCRIPTION_STATUSES:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="An active subscription is required",
            )

        if (
            SUBSCRIPTION_PLAN_LEVELS[subscription.plan]
            < SUBSCRIPTION_PLAN_LEVELS[required_plan]
        ):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail=f"The {required_plan.value} plan is required",
            )

        return subscription

    return dependency
